### ISSUE TITLE
Page flip option for E-Ink devices

### ISSUE THREAD

**Issue by tuomas2** [AndBible/and-bible#282]
From user via email:

> Can you add a page flip option for your Bible app to be more accommodating for those wanting to use this app from an Android E-Ink device? 
> 
> In order to be (properly) usable in an E-Ink device, page flip would have to be added as a selectable option to use instead of continuous scrolling. The page flip would need to have either no animations, or a "fade out/fade in" type of animation. Any sliding/moving animations (including the present continuous scrolling) create "ghosting" with E-Ink screens, and do not display well. Compatibility with any types of animations is very limited on E-Ink devices, and can make the user experience impractical.
> 
> If you could add an option for page flip in the future, this would allow your app to be usable on E-Ink devices. 
> So far, the only apps that are usable for Bible reading are Android Kindle and Kobo apps, which have an option to disable animations on page flips. Unfortunately, this means that I cannot effectively do Bible study on my e-reader.
> 
> If you have any questions about my suggestion, feel free to let me know.
>

**Comment by alerque** [comment]
I just came around to request this myself. I recently "upgraded" my eink device from an old Kindle to and Android based e-ink tablet (Boox Nova Air). My goal was mostly to break outside of the Amazon ecosystem so I could more easily load my own books, but the high degree of app functionality came as a surprise. This device is not Google Play certified, but I had no trouble loading up F-Droid and installing AndBible! It actually works great already and is quickly becoming a large chunk of my Bible reading.

This particular device has some features that make the scrolling interface usable, in particular it does a fast draw during scroll and then does a full repaint after the motion stops to get rid of the ghosting effects. This means I can already navigate and read the text, but the repaint experience is much less pleasant than using an ebook reader with a page based interface. If a page based view with next/previous page navigation instead of scroll could be added as an optional reading experience that would make a fantastic improvement.

**Comment by tuomas2** [comment]
#1126 related

**Comment by tuomas2** [comment]
does it have volume buttons? those work for scroll up/down

**Comment by alerque** [comment]
My edition has two half-way measures along those lines. There is a touch screen gesture for volume control, and the optional case adds hardware volume buttons.

Unfortunately scrolling (even a pageup/down chunk like the volume buttons do) is not a viable substitute for page-based layout. The scroll effects still screw up the e-ink display and make it difficult to track where you are.

**Comment by tuomas2** [comment]
I did not mean it as proper solution, but it might be still better than nothing if it would work. 

But yes, this is good request, I think it needs to be implemented. And should not be very difficult even. But requires some new UI options with text, so -> 4.1.

**Comment by tuomas2** [comment]
User would like to get buttons also like this: 

https://support.andbible.org/scp/tickets.php?id=1223

![Screenshot_20220118-094445-791](https://user-images.githubusercontent.com/5811789/150170701-e9fae60c-89c8-4606-a475-fd76437f7b15.png)

**Comment by agrogers** [comment]
We have the top buttons already don't we?

**Comment by tuomas2** [comment]
yes

**Comment by shumafuk** [comment]
I'd like to have this feature as well. Reading on Onyx Boox ereader is very uncomfortable with scrolling.

**Comment by alerque** [comment]
@tuomas2 I don't think that screenshot shows what we are after, those look like chapter navigation links. As evidenced by the half visible line that is not a paged interface, it is a scrolled one. Links to scroll a screen full would be better than nothing, but not at all as effective as actually rendering the text layout with a paged interface.

**Comment by agrogers** [comment]
I have only just started using a paged ebook on my phone and was surprised
how I liked it. Maybe it is something that non-ereaders would also use.

**Comment by alerque** [comment]
That's a good point that hasn't been mentioned yet: paged interfaces just make for a good reading experience even on non e-ink displays. It is harder to implement than scrolling, especially when the content is more variable (i.e. the web), but for plain text content without images or other significant layout issues that make line calculations hard it shouldn't be too hard to implement. I have to imagine there is some Android widgetry for paged canvases.

If/when this is implemented it shouldn't be gated behind an e-ink device detection, it should be available as a first class option for regular phones/tablets too.

**Comment by tuomas2** [comment]
> That's a good point that hasn't been mentioned yet: paged interfaces just make for a good reading experience even on non e-ink displays. It is harder to implement than scrolling, especially when the content is more variable (i.e. the web), but for plain text content without images or other significant layout issues that make line calculations hard it shouldn't be too hard to implement. I have to imagine there is some Android widgetry for paged canvases.

This is webview side stuff (not Android). 

> If/when this is implemented it shouldn't be gated behind an e-ink device detection, it should be available as a first class option for regular phones/tablets too.

I agree. Optional feature, adjusted with a setting.

**Comment by tuomas2** [comment]
Page flip implemented in #3354 by tapping the margin (if it is visible).

**Comment by tuomas2** [comment]
Implemented in #3354

**Comment by alerque** [comment]
I was able to get this to work, but the UX leaves a bit to be desired.

1. It is hard to discover. Even having seem this issue and having translated the UI (so I know I've seen all the strings) it still took me a while to find because it wasn't apparent that the text settings (not the app settings) were where this was, and after it was enabled it still didn't work at first because they were too small to hit. Typically mine are set at 0, and 6–8mm seems to be about the minimum to hit reliably on my device. 10mm is ever more reliable.

1. But that brings another issue: that cuts into the UI space *a lot* and gets a lot less text on the screen. More of an issue for smaller screens than larger I'm sure, but an issue non the less.

How about a setting where the swipe left/right action which currently skips by chapter changes to be a page-full scroll instead?

**Comment by tuomas2** [comment]
Good points. I can make setting for swipe gesture in addition to this. Hope it does not trigger full screen refresh because it is a swipe gesture (device might intepret it as scrolling the screen).

**Comment by tuomas2** [comment]
Implemented that.

**Comment by tuomas2** [comment]
Should there be a tip about next/prev page action in margin size setting?

**Comment by tuomas2** [comment]
added tip: "Tip: tap margin area to jmp one page up / down"

**Comment by alerque** [comment]
Thanks for that setting! I love it. After loading 837-5 I reset my margins to 0, changed the swipe action, and the navigation is great. Even with my device in speed mode the page-wise navigation redraws much cleaner than the old sliding scroll (which always meant artifacts) and doesn't even blink most of the time.
